crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+    fn set(&mut self, idx: usize) -> bool {
+        let n = idx / usize::BITS as usize;
+        let s = idx % usize::BITS as usize;
+        if (self.bits[n] & (1 << s)) == 0 {
+            self.bits[n] |= 1 << s;
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Returns an iterator of all unset indices.
+    fn unset(&self) -> impl Iterator<Item = usize> + '_ {
+        let mut index = 0;
+        let mut shift = 0;
+        let last_max_shift = self.capacity % usize::BITS as usize;
+        iter::from_fn(move || loop {
+            if shift >= usize::BITS as usize {
+                shift = 0;
+                index += 1;
+            }
+            if self.bits.len() <= index || (index >= self.bits.len() - 1 && shift >= last_max_shift)
+            {
+                return None;
+            }
+
+            let is_set = (self.bits[index] & (1 << shift)) != 0;
+            shift += 1;
+            if !is_set {
+                return Some((index * usize::BITS as usize) + shift - 1);
+            }
+        })
+    }
+}
+
+/// Set bits in `reached_map` for all blocks that are reached in `blocks`
+/// starting with block at index `idx`.
+fn mark_reached(reached_map: &mut Bitmap, blocks: &[BasicBlock<'_>], mut idx: BlockIndex) {
+    loop {
+        let block = &blocks[idx];
+        if reached_map.set(idx) {
+            return; // Block already visited, no needed to do it again.
+        }
+
+        match &block.next {
+            NextBlock::Always(next) => idx = *next,
+            NextBlock::If {
+                condition,
+                next,
+                orelse,
+            } => {
+                match taken(condition) {
+                    Some(true) => idx = *next,    // Always taken.
+                    Some(false) => idx = *orelse, // Never taken.
+                    None => {
+                        // Don't know, both branches might be taken.
+                        idx = *next;
+                        mark_reached(reached_map, blocks, *orelse);
+                    }
+                }
+            }
+            NextBlock::Terminate => return,
+        }
+    }
+}
+
+/// Determines if `condition` is taken.
+/// Returns `Some(true)` if the condition is always true, e.g. `if True`, same
+/// with `Some(false)` if it's never taken. If it can't be determined it returns
+/// `None`, e.g. `If i == 100`.
+fn taken(condition: &Condition) -> Option<bool> {
+    // TODO: add more cases to this where we can determine a condition
+    // statically. For now we only consider constant booleans.
+    match condition {
+        Condition::Test(expr) => match expr {
+            Expr::Constant(constant) => constant.value.as_bool().copied(),
+            _ => None,
+        },
+        Condition::Iterator(_) => None,
+        Condition::Match { .. } => None,
+    }
+}
+
+/// Collection of basic block.
+#[derive(Debug, PartialEq)]
+struct BasicBlocks<'stmt> {
```

Comment:
Not necessarily for this PR but I think it would make sense to extract `BasicBlock`s because they are useful for all sort of control flow computation. For example, we could use it to detect uninitialized/potentially uninitialized variable reads.

Reply:
Indeed, I didn't quite know where to put them, so I though I just put them in the rule file for now, makes avoiding merge conflicts a whole lot easier as well.